### PR TITLE
Debugger

### PR DESCRIPTION
**Pull request by mirceacaprioru**
Debugger implementation. The plugin is functional with support for multiple debugger windows.

### CODE REVIEW

File: src/regmapparser.cpp [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+		QDomElement addr = node.firstChildElement("Address");
+
+		uint hexNode = addr.text().split("0x")[1].toUInt(&status, 16);
+		uint hexAddress = address.split("0x")[1].toUInt(&status, 16);
+
+		if (hexAddress > lastAddress) {
+			return &lastNode;
+		}
+
+		//	if (registers.size() == (i+1))
```
Comment (dNechita) [inline src/regmapparser.cpp]: needs code cleanup

File: src/tool.cpp [hunk 2]
```diff
@@ -32,7 +32,7 @@ Tool::Tool(struct iio_context *ctx, QPushButton *runButton,
 	ctx(ctx), run_button(runButton), api(api),
 	name(name)
 {
-	run_button->parentWidget()->setDisabled(false);
+        run_button->parentWidget()->setDisabled(false);
```
Comment (dNechita) [inline src/tool.cpp]: you should use tabs instead of spaces.

File: src/tool_launcher.cpp [hunk 3]
```diff
@@ -444,6 +453,9 @@ ToolLauncher::~ToolLauncher()
 	delete tl_api;
 	delete ui;
 
+
+
+
```
Comment (dNechita) [inline src/tool_launcher.cpp]: lots of empty lines here

File: src/tool_launcher.hpp [hunk 4]
```diff
@@ -46,6 +46,7 @@
 #include "digitalio.hpp"
 #include "menuoption.h"
 #include "detachedWindow.hpp"
+#include "debugger.h"
```
Comment (dNechita) [inline src/tool_launcher.hpp]: this can go in the .cpp file

File: src/bitfieldwidget.cpp [hunk 5]
```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright 2016 Analog Devices, Inc.
```
Comment (dNechita) [inline src/bitfieldwidget.cpp]: 2017

**Comment by rgetz** [comment]
Can you post what the screen looks like - I don't think this was included in any of the mockups.

**Comment by mirceacaprioru** [comment]
image
This is a screenshot of the debugger.

**Comment by rgetz** [comment]
I think this has the functionality we are looking for, but maybe not the usibility. 

Have a look at:
http://swdownloads.analog.com/ACE/ACE_User_Manual_rev3.pdf
Section 3.8

https://mcuoneclipse.files.wordpress.com/2017/07/embsys.png

https://www.slideshare.net/mobile/valentin_ciocoi/exploring-so-cperipheralswitheclipseide

I will try to find some others, and Andrei should work with the designer to work on some mock-ups.

**Comment by mhennerich** [comment]
Thanks for the review and comments.
The debug view requires a force enable and is typically not available to regular Scopy users.
 
The current implementation of the debug view was lengthy disused during the weekly calls.
Current priority is on release. We can incorporate further requirements once we open Scopy for general purpose IIO usage. Nevertheless creating mock-ups for increased usability now sounds good.

**Pull request merged**